### ISSUE TITLE
Security contact for vulnerability report

### ISSUE THREAD

**Issue by RFCResearch** [yuzu-emu/yuzu#11871]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Affected Build(s)

N/A

### Description of Issue

Hello, 

I found multiple security issues in Yuzu that could be chained to execute code outside the emulator when running a malicious homebrew. Could you create a [Security Advisory](https://docs.github.com/en/code-security/security-advisories/working-with-repository-security-advisories/creating-a-repository-security-advisory) to share the details in private?

Thanks

### Expected Behavior

N/A

### Reproduction Steps

N/A

### Log File

N/A

### System Configuration

N/A

**Comment by Kelebek1** [comment]
Yuzu isn't designed to, or trying to be, secure.

**Comment by liamwhite** [comment]
Please see https://github.com/yuzu-emu/yuzu/issues/8547#issuecomment-1192036511

**Comment by Kelebek1** [comment]
I mean, I don't think these spambots are going to respond.

**Comment by RFCResearch** [comment]
Hi @Kelebek1, this is not related to Dynarmic. As far as I know, you should not be able to write/read arbitrary memory when going through Dynarmic as they are hooked to emulate the memory of the console

This is a chain of issues that allow escaping from the emulator into the user's computer. If Yuzu is expected to allow the compromise of your computer when executing homebrew, then users should at least be warned about it, as the expectation from users is definitely that Yuzu should **not allow complete access to your computer to software executing inside it**

I already tried to contact bunnei with no response. If you are okay with it, I can post the details (without a PoC) in this issue or create a new issue per vulnerability

**Comment by FernandoS27** [comment]
> Hi @Kelebek1, this is not related to Dynarmic. As far as I know, you should not be able to write/read arbitrary memory when going through Dynarmic as they are hooked to emulate the memory of the console
> 
> This is a chain of issues that allow escaping from the emulator into the user's computer. If Yuzu is expected to allow the compromise of your computer when executing homebrew, then users should at least be warned about it, as the expectation from users is definitely that Yuzu should **not allow complete access to your computer to software executing inside it**
> 
> I already tried to contact bunnei with no response. If you are okay with it, I can post the details (without a PoC) in this issue or create a new issue per vulnerability

we are discussing ways to mitigate this without affecting performance.

**Comment by liamwhite** [comment]
@RFCResearch in my opinion, you can document the issue publicly.

Within reason, we want to be able to prevent guest code from escaping into the host, but there are already many known avenues where this is possible due to design for performance:
- Signatures of guest code are not verified, to enable modifications
- Filesystem accesses within savedata are translated to direct host directory accesses, instead of using the savedata container format
- Guest programs can directory open the SD card filesystem as a directory and traverse the host this way
- Dynarmic does not extensively validate pointer accesses and this can be used to overwrite emulator memory (though there is option to enable this)

If you have found a new one, I am interested to hear it.

**Comment by RFCResearch** [comment]
There is another way to gain code execution directly

`yuzu/src/core/hle/service/nvdrv/devices/nvhost_gpu.cpp` contains multiple functions that seem vulnerable to buffer overflow issues. I was able to use `GetClientData` to leak pointers to bypass ASLR

```c++
NvResult nvhost_gpu::GetClientData(std::span<const u8> input, std::span<u8> output) {
    LOG_ERROR(Service_NVDRV, "called");

    IoctlClientData params{};
    std::memcpy(&params, input.data(), input.size()); // We control the size and the content of the buffer
    params.data = user_data;
    LOG_ERROR(Service_NVDRV, "Going to copy size={:X} from={} at={}", output.size(), fmt::ptr(&params),fmt::ptr(output.data()));
    std::memcpy(output.data(), &params, output.size()); // We also control the size of the buffer, so we can copy data past the buffer and leak pointers
    return NvResult::Success;
}
```

And then used `SetClientData`  to gain control of the stack

```c++
NvResult nvhost_gpu::SetClientData(std::span<const u8> input, std::span<u8> output) {
    LOG_DEBUG(Service_NVDRV, "called");

    IoctlClientData params{};
    std::memcpy(&params, input.data(), input.size()); // Again, we control the buffer and the size
    user_data = params.data;
    return NvResult::Success;
}
```

With these two issues I was able to escape from the emulator with a ROP chain and execute code on a Windows machine.

Example code for the leak
```c++
    u32 nvhostFd = 0;
    nvOpen(&nvhostFd, "/dev/nvhost-gpu");

    // Leak Yuzu Address
    unsigned char leakBuffer[500];
    memset(&leakBuffer, 0, sizeof(leakBuffer));
    // Modified version fo nvIoctl that allows setting the size of the read and write buffers manually
    nvIoctl_exploit(nvhostFd, _NV_IOWR(0x47, 0x15, leakBuffer), &leakBuffer, 10, 500);
```

On the other hand there is also a heap overflow located at `PollImpl`

```c++
std::pair<s32, Errno> BSD::PollImpl(std::vector<u8>& write_buffer, std::span<const u8> read_buffer,
                                    s32 nfds, s32 timeout) {
    if (write_buffer.size() < nfds * sizeof(PollFD)) {
        return {-1, Errno::INVAL};
    }

    if (nfds == 0) {
        // When no entries are provided, -1 is returned with errno zero
        return {-1, Errno::SUCCESS};
    }

    const size_t length = std::min(read_buffer.size(), write_buffer.size());
    std::vector<PollFD> fds(nfds); // Argument controlled by the user, we can set it to 1 for example
    std::memcpy(fds.data(), read_buffer.data(), length); // We control length, so we copy more than the buffer is expecting, as we control the size of the vector used as destination

    if (timeout >= 0) {
        const s64 seconds = timeout / 1000;
        const u64 nanoseconds = 1'000'000 * (static_cast<u64>(timeout) % 1000);

        if (seconds < 0) {
            return {-1, Errno::INVAL};
        }
        if (nanoseconds > 999'999'999) {
            return {-1, Errno::INVAL};
        }
    } else if (timeout != -1) {
        return {-1, Errno::INVAL};
    }
```

This issue also seems exploitable with an information leak. With specific sizes you can get a chunk located between memory structures so you could potentially overwrite pointers of important structures to locations in your program to control the behavior of Yuzu (or exploit libc issues, depending of the target)

Finally, about being able to access the filesystem, just in case you are not aware, it also allows direct execution of code outside the emulator on Linux by opening `/proc/self/mem` to overwrite the data section of Yuzu from a homebrew

**Comment by liamwhite** [comment]
Thanks for the report. If you have more, we will be happy to fix them, of course.

**Comment by RFCResearch** [comment]
Thanks for the quick response. If you search for `std::memcpy(&params, input.data(), input.size());` and `std::memcpy(output.data(), &params, output.size());` (probably better with a regex instead of `&params`) you will see multiple cases of the same issue in files at `yuzu/src/core/hle/service/nvdrv/devices/`. I haven't tried to trigger them all, but they if they follow the same pattern, the user controls the input too.

For example
```c++
NvResult nvhost_gpu::SetNVMAPfd(std::span<const u8> input, std::span<u8> output) {
    IoctlSetNvmapFD params{};
    std::memcpy(&params, input.data(), input.size()); // We control the input size
    LOG_DEBUG(Service_NVDRV, "called, fd={}", params.nvmap_fd);

    nvmap_fd = params.nvmap_fd;
    return NvResult::Success;
}
```

It might be useful to create some semgrep rules and make them part of the CI to detect these issues early

**Comment by liamwhite** [comment]
Indeed, the entirety of nvservices module is filled with these errors.

**Comment by liamwhite** [comment]
@RFCResearch fixed all of these now in the linked pull request by adding a proper command serialization system for nvservices. A review would be appreciated, thanks!

**Comment by RFCResearch** [comment]
@liamwhite I took a look at it looks good. Very nice patch
